qutip/solver/heom/bofin_solvers.py

```diff
@@ -745,6 +756,16 @@ def _grad_prev_bosonic(self, he_n, k):
             )
         return op
 
+    def _get_signs(self, n_excite, n_excite_before_m):
```

Comment:
Extracting the signs out into this function destroys the readability of the logic. Functions should not be use to merely reduce repetitive, but rather to abstract out concepts that clarify understanding.

This function does the opposite -- it makes the logic *harder* to understand.

I'll suggest an alternative below.